File: vscode/src/core/logs/logParser.frameworkSample.test.ts
import * as assert from 'assert';
import { isError, isNetwork } from './logCategory';
import { FRAMEWORK_SAMPLE } from './logParser.fixtures';
import { parseLog } from './logParser';
import { NetworkEntry, StandardLine } from './logModel';

/**
 * Contract guard: parses the *exact* NDJSON lines emitted by `nylo_support`'s `NyFileLogger`
 * (captured from its Phase A test run) so a future change on either side that breaks the shared
 * schema fails here. Verifies microsecond timestamps and superset network fields are tolerated.
 * Mirror of the Kotlin `LogParserFrameworkSampleTest` — keep the two in sync.
 */
describe('logParser framework sample', () => {
  it('parses the real framework NDJSON output', () => {
    const doc = parseLog(FRAMEWORK_SAMPLE);
    assert.strictEqual(doc.sessions.length, 1);
    const s = doc.sessions[0];

    assert.strictEqual(s.tag, 'idrxkz');
    assert.strictEqual(s.fullId, '2026-06-26T14-17-24-idrxkz');
    assert.strictEqual(s.version, 'nylo_framework v7.1.24');
    // Microsecond fraction (.223296) survives into the sort key; display drops it.
    const base = Date.UTC(2026, 5, 26, 14, 17, 24);
    assert.ok(s.started! > base + 223 && s.started! < base + 224);
    assert.strictEqual(s.startedRaw, '2026-06-26 14:17:24');

    const stds = s.entries.filter((e): e is StandardLine => e.entryType === 'standard');
    assert.strictEqual(stds.length, 3); // debug, error(+context+stack), console
    const errorLine = stds.filter((e) => e.level === 'error')[0];
    assert.ok(errorLine.raw.includes('Failed to load user'));
    assert.ok(errorLine.raw.includes('"retry":true')); // context rendered inline
    assert.ok(errorLine.raw.includes('\n#0 main')); // stack appended
    assert.strictEqual(stds.filter((e) => e.message.includes('raw console line'))[0].level, null);

    const nets = s.entries.filter((e): e is NetworkEntry => e.entryType === 'network');
    assert.strictEqual(nets.length, 3);
    assert.strictEqual(nets[0].netKind, 'request');
    assert.strictEqual(nets[0].requestId, '236f0833');
    assert.ok(nets[0].summary.includes('[PUT]'));
    assert.ok(nets[0].raw.includes('Ada')); // body folded in; empty headers{} skipped
    assert.ok(!nets[0].raw.includes('headers'));
    assert.strictEqual(nets[1].statusCode, 200);
    assert.ok(nets[1].raw.includes('"name": "Ada"')); // response body pretty-printed
    assert.strictEqual(nets[2].netKind, 'error');
    assert.strictEqual(nets[2].statusCode, 500);

    // Categorization: the error log line and the net error both land in ERRORS.
    assert.ok(isError(errorLine));
    assert.ok(isError(nets[2]));
    assert.ok(isNetwork(nets[0]));
  });
});

File: intellij/src/test/kotlin/dev/nylo/plugin/logs/parse/LogParserFrameworkSampleTest.kt
package dev.nylo.plugin.logs.parse

import dev.nylo.plugin.logs.model.LogCategorizer
import dev.nylo.plugin.logs.model.NetworkEntry
import dev.nylo.plugin.logs.model.StandardLine
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDateTime

/**
 * Contract guard: parses the *exact* NDJSON lines emitted by `nylo_support`'s `NyFileLogger`
 * (captured from its Phase A test run) so a future change on either side that breaks the shared
 * schema fails here. Verifies microsecond timestamps and superset network fields are tolerated.
 */
class LogParserFrameworkSampleTest {

    private val sample = listOf(
        """{"t":"session","id":"2026-06-26T14-17-24-idrxkz","started":"2026-06-26T14:17:24.223296","platform":"macos Version 26.5.1 (Build 25F80)","app":"Pretalk 1.0.0","version":"nylo_framework v7.1.24","env":"developing"}""",
        """{"t":"log","ts":"2026-06-26T10:32:59.123","session":"idrxkz","level":"debug","msg":"[AppProvider] Booting"}""",
        """{"t":"log","ts":"2026-06-26T10:33:00.001","session":"idrxkz","level":"error","msg":"Failed to load user","context":{"id":123,"retry":true},"stack":"#0 main (file.dart:1:1)\n#1 x (y:2:2)"}""",
        """{"t":"console","ts":"2026-06-26T14:17:24.229440","session":"idrxkz","msg":"flutter: raw console line"}""",
        """{"t":"net","session":"idrxkz","ts":"2026-06-26T14:17:24.238011","kind":"request","requestId":"236f0833","method":"PUT","uri":"https://api.example.com/users/5","headers":{},"contentType":null,"responseType":"ResponseType.json","body":{"name":"Ada"}}""",
        """{"t":"net","session":"idrxkz","ts":"2026-06-26T14:17:24.300000","kind":"response","requestId":"236f0833","method":"PUT","uri":"https://api.example.com/users/5","statusCode":200,"statusMessage":"OK","responseTimeMs":5,"payloadSizeBytes":18,"payloadSize":"18 B","data":{"id":5,"name":"Ada"}}""",
        """{"t":"net","session":"idrxkz","ts":"2026-06-26T14:17:24.400000","kind":"error","requestId":"236f0833","errorType":"DioExceptionType.badResponse","method":"PUT","uri":"https://api.example.com/users/5","responseTimeMs":7,"payloadSizeBytes":15,"payloadSize":"15 B","statusCode":500,"statusMessage":"Server Error","data":{"error":"kaboom"},"message":"Http status error [500]"}""",
    ).joinToString("\n")

    @Test
    fun `parses the real framework NDJSON output`() {
        val s = LogParser.parse(sample).sessions.single()

        assertEquals("idrxkz", s.tag)
        assertEquals("2026-06-26T14-17-24-idrxkz", s.fullId)
        assertEquals("nylo_framework v7.1.24", s.version)
        // Microsecond fraction (.223296) parses; display drops it.
        assertEquals(LocalDateTime.of(2026, 6, 26, 14, 17, 24, 223_296_000), s.started)
        assertEquals("2026-06-26 14:17:24", s.startedRaw)

        val stds = s.entries.filterIsInstance<StandardLine>()
        assertEquals(3, stds.size) // debug, error(+context+stack), console
        val errorLine = stds.single { it.level == "error" }
        assertTrue(errorLine.raw.contains("Failed to load user"))
        assertTrue(errorLine.raw.contains("\"retry\":true")) // context rendered inline
        assertTrue(errorLine.raw.contains("\n#0 main")) // stack appended
        assertNull(stds.single { it.message.contains("raw console line") }.level) // console has no level

        val nets = s.entries.filterIsInstance<NetworkEntry>()
        assertEquals(3, nets.size)
        assertEquals(NetworkEntry.Kind.REQUEST, nets[0].kind)
        assertEquals("236f0833", nets[0].requestId)
        assertTrue(nets[0].summary.contains("[PUT]"))
        assertTrue(nets[0].raw.contains("Ada")) // body folded in; empty headers{} skipped
        assertTrue(nets[0].raw.contains("headers").not())
        assertEquals(200, nets[1].statusCode)
        assertTrue(nets[1].raw.contains("\"name\": \"Ada\"")) // response body pretty-printed
        assertEquals(NetworkEntry.Kind.ERROR, nets[2].kind)
        assertEquals(500, nets[2].statusCode)

        // Categorization: the error log line and the net error both land in ERRORS.
        assertTrue(LogCategorizer.isError(errorLine))
        assertTrue(LogCategorizer.isError(nets[2]))
        assertTrue(LogCategorizer.isNetwork(nets[0]))
    }
}
